giles-eco/src/main/java/edu/asu/diging/gilesecosystem/web/tokens/impl/IntrospectTokenService.java

```diff
@@ -0,0 +1,157 @@
+
+        IRegisteredApp app = appsManager.getApp(appId);
+
+        if (app.getProviderClientId() == null || app.getProviderClientId().isEmpty()) {
+            throw new AppMisconfigurationException("No provider client id has been registered for your app.");
+        }
+
+        // use mitreid connect client with protected resource access
+        final String clientId = propertyManager.getProperty(Properties.MITREID_INTROSPECT_CLIENT_ID);
+        final String clientSecret = propertyManager.getProperty(Properties.MITREID_INTROSPECT_SECRET);
+
+        String introspectionUrl = propertyManager.getProperty(Properties.MITREID_INTROSPECT_URL);
+
+        if (introspectionUrl == null || introspectionUrl.isEmpty()) {
+            throw new AppMisconfigurationException("Unable to load server URL");
+        }
+
+        MultiValueMap<String, String> form = new LinkedMultiValueMap<>();
+
+        RestTemplate restTemplate = new RestTemplate(factory) {
+            @Override
+            protected ClientHttpRequest createRequest(URI url, HttpMethod method) throws IOException {
+                ClientHttpRequest httpRequest = super.createRequest(url, method);
+                httpRequest.getHeaders().add("Authorization",
+                        String.format("Basic %s", Base64.encode(String.format("%s:%s", clientId, clientSecret))));
+                return httpRequest;
+            }
+        };
+
+        form.add("token", accessToken);
+        String validatedToken = null;
+        try {
+            validatedToken = restTemplate.postForObject(introspectionUrl, form, String.class);
+        } catch (RestClientException rce) {
+            logger.error("introspectToken", rce);
+            return null;
+        }
+
+        return parseValidatedToken(validatedToken);
+    }
+
+    private IApiTokenContents parseValidatedToken(String validatedToken) {
+
+        if (validatedToken == null) {
+            return null;
+        }
+
+        // parse the json
+        JsonElement jsonRoot = new JsonParser().parse(validatedToken);
+        if (!jsonRoot.isJsonObject()) {
+            return null; // didn't get a proper JSON object
+        }
+
+        JsonObject tokenResponse = jsonRoot.getAsJsonObject();
+
+        if (tokenResponse.get("error") != null) {
+            logger.error("Got an error back: " + tokenResponse.get("error") + ", "
+                    + tokenResponse.get("error_description"));
+            return null;
+        }
+
+        if (!tokenResponse.get("active").getAsBoolean()) {
+            // non-valid token
+            logger.info("Server returned non-active token");
+            return null;
+        }
+
+        // validated token is not null and has no error, it is a valid
+        // access token
+        return updateTokenContents(tokenResponse);
+    }
+
+    private IApiTokenContents updateTokenContents(JsonObject tokenResponse) {
```

Comment:
it create tokenContents object and update it with values received from rest service response